File: tasks/commands/check.py
"""Consistency checks for task tree integrity."""

from __future__ import annotations

import json
import re
from pathlib import Path

import click
import networkx as nx
from rich.console import Console

from ..helpers import (
    CONTEXT_FILE,
    SESSIONS_FILE,
    find_missing_task_files,
    get_all_tasks,
)
from ..loader import TaskLoader

console = Console()

PHASE_ID_RE = re.compile(r"^P\d+$")
MILESTONE_ID_RE = re.compile(r"^P\d+\.M\d+$")
EPIC_ID_RE = re.compile(r"^P\d+\.M\d+\.E\d+$")
TASK_ID_RE = re.compile(r"^P\d+\.M\d+\.E\d+\.T\d+$")


def _add_finding(
    findings, level: str, code: str, message: str, location: str | None = None
):
    finding = {"level": level, "code": code, "message": message}
    if location:
        finding["location"] = location
    findings.append(finding)


def _validate_tree_files(tree, findings):
    tasks_root = Path(".tasks")
    root_index = tasks_root / "index.yaml"
    if not root_index.exists():
        _add_finding(
            findings, "error", "missing_root_index", "Missing .tasks/index.yaml"
        )

    for phase in tree.phases:
        phase_index = tasks_root / phase.path / "index.yaml"
        if not phase_index.exists():
            _add_finding(
                findings,
                "error",
                "missing_phase_index",
                f"Missing phase index for {phase.id}",
                str(phase_index),
            )

        for milestone in phase.milestones:
            milestone_index = tasks_root / phase.path / milestone.path / "index.yaml"
            if not milestone_index.exists():
                _add_finding(
                    findings,
                    "error",
                    "missing_milestone_index",
                    f"Missing milestone index for {milestone.id}",
                    str(milestone_index),
                )

            for epic in milestone.epics:
                epic_index = (
                    tasks_root / phase.path / milestone.path / epic.path / "index.yaml"
                )
                if not epic_index.exists():
                    _add_finding(
                        findings,
                        "error",
                        "missing_epic_index",
                        f"Missing epic index for {epic.id}",
                        str(epic_index),
                    )

    for task in find_missing_task_files(tree):
        _add_finding(
            findings,
            "error",
            "missing_task_file",
            f"Task file missing for {task.id}",
            f".tasks/{task.file}",
        )


def _resolve_milestone_dep_id(dep_id: str, phase_id: str | None) -> str | None:
    dep = dep_id.strip()
    if not dep:
        return None
    if "." in dep:
        return dep
    if phase_id:
        return f"{phase_id}.{dep}"
    return dep


def _resolve_epic_dep_id(dep_id: str, milestone_id: str | None) -> str | None:
    dep = dep_id.strip()
    if not dep:
        return None
    if "." in dep:
        return dep
    if milestone_id:
        return f"{milestone_id}.{dep}"
    return dep


def _validate_ids_and_dependencies(tree, findings):
    all_tasks = get_all_tasks(tree)

    phase_ids = set()
    milestone_ids = set()
    epic_ids = set()
    task_ids = set()

    for phase in tree.phases:
        if not PHASE_ID_RE.match(phase.id):
            _add_finding(
                findings,
                "error",
                "invalid_phase_id",
                f"Invalid phase ID format: {phase.id}",
                phase.id,
            )
        if phase.id in phase_ids:
            _add_finding(
                findings,
                "error",
                "duplicate_phase_id",
                f"Duplicate phase ID: {phase.id}",
                phase.id,
            )
        phase_ids.add(phase.id)

        for dep in phase.depends_on:
            if dep == phase.id:
                _add_finding(
                    findings,
                    "error",
                    "self_dependency_phase",
                    f"Phase depends on itself: {phase.id}",
                    phase.id,
                )
            elif dep not in phase_ids and not any(p.id == dep for p in tree.phases):
                _add_finding(
                    findings,
                    "error",
                    "missing_phase_dependency",
                    f"Phase dependency not found: {dep}",
                    phase.id,
                )

        for milestone in phase.milestones:
            if not MILESTONE_ID_RE.match(milestone.id):
                _add_finding(
                    findings,
                    "error",
                    "invalid_milestone_id",
                    f"Invalid milestone ID format: {milestone.id}",
                    milestone.id,
                )
            if milestone.id in milestone_ids:
                _add_finding(
                    findings,
                    "error",
                    "duplicate_milestone_id",
                    f"Duplicate milestone ID: {milestone.id}",
                    milestone.id,
                )
            milestone_ids.add(milestone.id)

            if milestone.phase_id != phase.id:
                _add_finding(
                    findings,
                    "error",
                    "milestone_parent_mismatch",
                    f"Milestone {milestone.id} is attached to phase {phase.id} but reports phase_id={milestone.phase_id}",
                    milestone.id,
                )

            for dep in milestone.depends_on:
                dep_resolved = _resolve_milestone_dep_id(dep, phase.id)
                if dep_resolved == milestone.id:
                    _add_finding(
                        findings,
                        "error",
                        "self_dependency_milestone",
                        f"Milestone depends on itself: {milestone.id}",
                        milestone.id,
                    )
                elif dep_resolved and not tree.find_milestone(dep_resolved):
                    _add_finding(
                        findings,
                        "error",
                        "missing_milestone_dependency",
                        f"Milestone dependency not found: {dep}",
                        milestone.id,
                    )

            for epic in milestone.epics:
                if not EPIC_ID_RE.match(epic.id):
                    _add_finding(
                        findings,
                        "error",
                        "invalid_epic_id",
                        f"Invalid epic ID format: {epic.id}",
                        epic.id,
                    )
                if epic.id in epic_ids:
                    _add_finding(
                        findings,
                        "error",
                        "duplicate_epic_id",
                        f"Duplicate epic ID: {epic.id}",
                        epic.id,
                    )
                epic_ids.add(epic.id)

                if epic.phase_id != phase.id:
                    _add_finding(
                        findings,
                        "error",
                        "epic_phase_mismatch",
                        f"Epic {epic.id} has phase_id={epic.phase_id}, expected {phase.id}",
                        epic.id,
                    )
                if epic.milestone_id != milestone.id:
                    _add_finding(
                        findings,
                        "error",
                        "epic_milestone_mismatch",
                        f"Epic {epic.id} has milestone_id={epic.milestone_id}, expected {milestone.id}",
                        epic.id,
                    )

                for dep in epic.depends_on:
                    dep_resolved = _resolve_epic_dep_id(dep, milestone.id)
                    if dep_resolved == epic.id:
                        _add_finding(
                            findings,
                            "error",
                            "self_dependency_epic",
                            f"Epic depends on itself: {epic.id}",
                            epic.id,
                        )
                    elif dep_resolved and not tree.find_epic(dep_resolved):
                        _add_finding(
                            findings,
                            "error",
                            "missing_epic_dependency",
                            f"Epic dependency not found: {dep}",
                            epic.id,
                        )

                expected_prefix = f".tasks/{phase.path}/{milestone.path}/{epic.path}/"
                for task in epic.tasks:
                    if not TASK_ID_RE.match(task.id):
                        _add_finding(
                            findings,
                            "error",
                            "invalid_task_id",
                            f"Invalid task ID format: {task.id}",
                            task.id,
                        )
                    if task.id in task_ids:
                        _add_finding(
                            findings,
                            "error",
                            "duplicate_task_id",
                            f"Duplicate task ID: {task.id}",
                            task.id,
                        )
                    task_ids.add(task.id)

                    if task.epic_id != epic.id:
                        _add_finding(
                            findings,
                            "error",
                            "task_epic_mismatch",
                            f"Task {task.id} has epic_id={task.epic_id}, expected {epic.id}",
                            task.id,
                        )
                    if task.milestone_id != milestone.id:
                        _add_finding(
                            findings,
                            "error",
                            "task_milestone_mismatch",
                            f"Task {task.id} has milestone_id={task.milestone_id}, expected {milestone.id}",
                            task.id,
                        )
                    if task.phase_id != phase.id:
                        _add_finding(
                            findings,
                            "error",
                            "task_phase_mismatch",
                            f"Task {task.id} has phase_id={task.phase_id}, expected {phase.id}",
                            task.id,
                        )

                    if not task.id.startswith(f"{epic.id}."):
                        _add_finding(
                            findings,
                            "error",
                            "task_parent_path_mismatch",
                            f"Task {task.id} does not belong to parent epic {epic.id}",
                            task.id,
                        )

                    if not task.file.endswith(".todo"):
                        _add_finding(
                            findings,
                            "error",
                            "invalid_task_file_extension",
                            f"Task file must end with .todo: {task.file}",
                            task.id,
                        )

                    if not f".tasks/{task.file}".startswith(expected_prefix):
                        _add_finding(
                            findings,
                            "error",
                            "task_file_path_mismatch",
                            f"Task file path does not match parent hierarchy: .tasks/{task.file}",
                            task.id,
                        )

                    if task.estimate_hours == 0:
                        _add_finding(
                            findings,
                            "warning",
                            "zero_estimate_hours",
                            f"Task estimate must be positive, got 0: {task.id}",
                            task.id,
                        )

    for task in all_tasks:
        for dep in task.depends_on:
            if dep == task.id:
                _add_finding(
                    findings,
                    "error",
                    "self_dependency_task",
                    f"Task depends on itself: {task.id}",
                    task.id,
                )
            elif dep not in task_ids:
                _add_finding(
                    findings,
                    "error",
                    "missing_task_dependency",
                    f"Task dependency not found: {dep}",
                    task.id,
                )


def _append_graph_cycle_findings(
    graph, findings, code: str, message_prefix: str, location: str
):
    if nx.is_directed_acyclic_graph(graph):
        return
    cycles = list(nx.simple_cycles(graph))
    for cycle in cycles[:10]:
        if not cycle:
            continue
        chain = " -> ".join(cycle + [cycle[0]])
        _add_finding(findings, "error", code, f"{message_prefix}: {chain}", location)


def _validate_cycles(tree, findings):
    # Task dependency cycles
    task_graph = nx.DiGraph()
    all_tasks = get_all_tasks(tree)
    for task in all_tasks:
        task_graph.add_node(task.id)
    for phase in tree.phases:
        for milestone in phase.milestones:
            for epic in milestone.epics:
                for idx, task in enumerate(epic.tasks):
                    for dep in task.depends_on:
                        if task_graph.has_node(dep):
                            task_graph.add_edge(dep, task.id)
                    if not task.depends_on and idx > 0:
                        task_graph.add_edge(epic.tasks[idx - 1].id, task.id)
    _append_graph_cycle_findings(
        task_graph,
        findings,
        "task_dependency_cycle",
        "Task dependency cycle detected",
        "task_dependencies",
    )

    # Epic dependency cycles
    epic_graph = nx.DiGraph()
    for phase in tree.phases:
        for milestone in phase.milestones:
            for epic in milestone.epics:
                epic_graph.add_node(epic.id)
    for phase in tree.phases:
        for milestone in phase.milestones:
            for epic in milestone.epics:
                for dep in epic.depends_on:
                    dep_resolved = _resolve_epic_dep_id(dep, milestone.id)
                    if dep_resolved and epic_graph.has_node(dep_resolved):
                        epic_graph.add_edge(dep_resolved, epic.id)
    _append_graph_cycle_findings(
        epic_graph,
        findings,
        "epic_dependency_cycle",
        "Epic dependency cycle detected",
        "epic_dependencies",
    )

    # Milestone dependency cycles
    milestone_graph = nx.DiGraph()
    for phase in tree.phases:
        for milestone in phase.milestones:
            milestone_graph.add_node(milestone.id)
    for phase in tree.phases:
        for milestone in phase.milestones:
            for dep in milestone.depends_on:
                dep_resolved = _resolve_milestone_dep_id(dep, phase.id)
                if dep_resolved and milestone_graph.has_node(dep_resolved):
                    milestone_graph.add_edge(dep_resolved, milestone.id)
    _append_graph_cycle_findings(
        milestone_graph,
        findings,
        "milestone_dependency_cycle",
        "Milestone dependency cycle detected",
        "milestone_dependencies",
    )

    # Phase dependency cycles
    phase_graph = nx.DiGraph()
    for phase in tree.phases:
        phase_graph.add_node(phase.id)
    for phase in tree.phases:
        for dep in phase.depends_on:
            if phase_graph.has_node(dep):
                phase_graph.add_edge(dep, phase.id)
    _append_graph_cycle_findings(
        phase_graph,
        findings,
        "phase_dependency_cycle",
        "Phase dependency cycle detected",
        "phase_dependencies",
    )


def _validate_runtime_files(tree, findings):
    if CONTEXT_FILE.exists():
        try:
            import yaml

            with open(CONTEXT_FILE) as f:
                context = yaml.safe_load(f) or {}
            referenced = []
            current = context.get("current_task")
            primary = context.get("primary_task")
            additional = context.get("additional_tasks") or []
            siblings = context.get("sibling_tasks") or []
            if current:
                referenced.append(current)
            if primary:
                referenced.append(primary)
            referenced.extend(additional)
            referenced.extend(siblings)
            for task_id in referenced:
                if not tree.find_task(task_id):
                    _add_finding(
                        findings,
                        "warning",
                        "stale_context_task",
                        f"Context references missing task: {task_id}",
                        str(CONTEXT_FILE),
                    )
        except Exception as exc:
            _add_finding(
                findings,
                "warning",
                "context_parse_error",
                f"Could not parse context file: {exc}",
                str(CONTEXT_FILE),
            )

    if SESSIONS_FILE.exists():
        try:
            import yaml

            with open(SESSIONS_FILE) as f:
                sessions = yaml.safe_load(f) or {}
            for _agent, payload in sessions.items():
                task_id = (
                    payload.get("current_task") if isinstance(payload, dict) else None
                )
                if task_id and not tree.find_task(task_id):
                    _add_finding(
                        findings,
                        "warning",
                        "stale_session_task",
                        f"Session references missing task: {task_id}",
                        str(SESSIONS_FILE),
                    )
        except Exception as exc:
            _add_finding(
                findings,
                "warning",
                "sessions_parse_error",
                f"Could not parse sessions file: {exc}",
                str(SESSIONS_FILE),
            )


def run_checks(tasks_dir: str = ".tasks") -> dict:
    """Run consistency checks and return normalized findings."""
    findings = []

    loader = TaskLoader(tasks_dir)
    tree = loader.load()

    _validate_tree_files(tree, findings)
    _validate_ids_and_dependencies(tree, findings)
    _validate_cycles(tree, findings)
    _validate_runtime_files(tree, findings)

    errors = [f for f in findings if f["level"] == "error"]
    warnings = [f for f in findings if f["level"] == "warning"]
    return {
        "ok": len(errors) == 0,
        "errors": errors,
        "warnings": warnings,
        "summary": {
            "errors": len(errors),
            "warnings": len(warnings),
            "total": len(findings),
        },
    }


def _print_text_report(report: dict):
    errors = report["errors"]
    warnings = report["warnings"]
    summary = report["summary"]

    if not errors and not warnings:
        console.print("[green]✓ Consistency check passed with no issues.[/]")
        return

    console.print(
        f"[bold]Consistency check results:[/] {summary['errors']} error(s), {summary['warnings']} warning(s)"
    )

    if errors:
        console.print("\n[red]Errors:[/]")
        for item in errors:
            loc = f" ({item['location']})" if "location" in item else ""
            console.print(f"  - {item['code']}: {item['message']}{loc}")

    if warnings:
        console.print("\n[yellow]Warnings:[/]")
        for item in warnings:
            loc = f" ({item['location']})" if "location" in item else ""
            console.print(f"  - {item['code']}: {item['message']}{loc}")


@click.command()
@click.option("--json", "as_json", is_flag=True, help="Output report as JSON.")
@click.option(
    "--strict",
    is_flag=True,
    help="Treat warnings as failures (non-zero exit).",
)
def check(as_json, strict):
    """Check task tree consistency (files, dependencies, cycles, IDs)."""
    try:
        report = run_checks()
    except Exception as exc:
        raise click.ClickException(str(exc)) from exc

    if as_json:
        click.echo(json.dumps(report, indent=2))
    else:
        _print_text_report(report)

    should_fail = bool(report["errors"]) or (strict and bool(report["warnings"]))
    if should_fail:
        raise click.exceptions.Exit(1)


def register_commands(cli):
    """Register check command."""
    cli.add_command(check)

File: backlog_ts/src/check.ts
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { parse } from "yaml";
import { findEpic, findMilestone, findPhase, findTask, getAllTasks, isTaskFileMissing } from "./helpers";
import { TaskLoader } from "./loader";
import type { TaskTree } from "./models";
import { getDataDirName } from "./data_dir";

type Finding = {
  level: "error" | "warning";
  code: string;
  message: string;
  location?: string;
};

const PHASE_ID_RE = /^P\d+$/;
const MILESTONE_ID_RE = /^P\d+\.M\d+$/;
const EPIC_ID_RE = /^P\d+\.M\d+\.E\d+$/;
const TASK_ID_RE = /^P\d+\.M\d+\.E\d+\.T\d+$/;
const BUG_ID_RE = /^B\d+$/;

function addFinding(findings: Finding[], finding: Finding): void {
  findings.push(finding);
}

function resolveMilestoneDepId(depId: string, phaseId?: string): string {
  const dep = depId.trim();
  if (!dep) return dep;
  if (dep.includes(".")) return dep;
  return phaseId ? `${phaseId}.${dep}` : dep;
}

function resolveEpicDepId(depId: string, milestoneId?: string): string {
  const dep = depId.trim();
  if (!dep) return dep;
  if (dep.includes(".")) return dep;
  return milestoneId ? `${milestoneId}.${dep}` : dep;
}

function hasDirectedCycle(nodes: string[], edges: Array<[string, string]>): string[] | null {
  const adj = new Map<string, string[]>();
  for (const n of nodes) adj.set(n, []);
  for (const [a, b] of edges) {
    if (!adj.has(a)) adj.set(a, []);
    adj.get(a)!.push(b);
  }

  const visiting = new Set<string>();
  const visited = new Set<string>();
  const stack: string[] = [];

  const dfs = (n: string): string[] | null => {
    if (visiting.has(n)) {
      const idx = stack.indexOf(n);
      return idx >= 0 ? stack.slice(idx).concat(n) : [n, n];
    }
    if (visited.has(n)) return null;
    visiting.add(n);
    stack.push(n);
    for (const m of adj.get(n) ?? []) {
      const cyc = dfs(m);
      if (cyc) return cyc;
    }
    stack.pop();
    visiting.delete(n);
    visited.add(n);
    return null;
  };

  for (const n of nodes) {
    const cyc = dfs(n);
    if (cyc) return cyc;
  }
  return null;
}

function validateTreeFiles(tree: TaskTree, findings: Finding[]): void {
  const dataDir = getDataDirName();
  if (!existsSync(`${dataDir}/index.yaml`)) {
    addFinding(findings, {
      level: "error",
      code: "missing_root_index",
      message: `Missing ${dataDir}/index.yaml`,
    });
  }

  for (const phase of tree.phases) {
    const phaseIndex = `${dataDir}/${phase.path}/index.yaml`;
    if (!existsSync(phaseIndex)) {
      addFinding(findings, {
        level: "error",
        code: "missing_phase_index",
        message: `Missing phase index for ${phase.id}`,
        location: phaseIndex,
      });
    }

    for (const milestone of phase.milestones) {
      const milestoneIndex = `${dataDir}/${phase.path}/${milestone.path}/index.yaml`;
      if (!existsSync(milestoneIndex)) {
        addFinding(findings, {
          level: "error",
          code: "missing_milestone_index",
          message: `Missing milestone index for ${milestone.id}`,
          location: milestoneIndex,
        });
      }
      for (const epic of milestone.epics) {
        const epicIndex = `${dataDir}/${phase.path}/${milestone.path}/${epic.path}/index.yaml`;
        if (!existsSync(epicIndex)) {
          addFinding(findings, {
            level: "error",
            code: "missing_epic_index",
            message: `Missing epic index for ${epic.id}`,
            location: epicIndex,
          });
        }
      }
    }
  }

  for (const t of getAllTasks(tree)) {
    if (isTaskFileMissing(t)) {
      addFinding(findings, {
        level: "error",
        code: "missing_task_file",
        message: `Task file missing for ${t.id}`,
        location: `${dataDir}/${t.file}`,
      });
    }
  }
}

function validateIdsAndDependencies(tree: TaskTree, findings: Finding[]): void {
  const allTasks = getAllTasks(tree);
  const taskIds = new Set<string>();
  const phaseIds = new Set<string>();
  const milestoneIds = new Set<string>();
  const epicIds = new Set<string>();

  for (const phase of tree.phases) {
    if (!PHASE_ID_RE.test(phase.id)) addFinding(findings, { level: "error", code: "invalid_phase_id", message: `Invalid phase ID format: ${phase.id}`, location: phase.id });
    if (phaseIds.has(phase.id)) addFinding(findings, { level: "error", code: "duplicate_phase_id", message: `Duplicate phase ID: ${phase.id}`, location: phase.id });
    phaseIds.add(phase.id);

    for (const dep of phase.dependsOn) {
      if (dep === phase.id) addFinding(findings, { level: "error", code: "self_dependency_phase", message: `Phase depends on itself: ${phase.id}`, location: phase.id });
      else if (!findPhase(tree, dep)) addFinding(findings, { level: "error", code: "missing_phase_dependency", message: `Phase dependency not found: ${dep}`, location: phase.id });
    }

    for (const milestone of phase.milestones) {
      if (!MILESTONE_ID_RE.test(milestone.id)) addFinding(findings, { level: "error", code: "invalid_milestone_id", message: `Invalid milestone ID format: ${milestone.id}`, location: milestone.id });
      if (milestoneIds.has(milestone.id)) addFinding(findings, { level: "error", code: "duplicate_milestone_id", message: `Duplicate milestone ID: ${milestone.id}`, location: milestone.id });
      milestoneIds.add(milestone.id);

      for (const dep of milestone.dependsOn) {
        const resolved = resolveMilestoneDepId(dep, phase.id);
        if (resolved === milestone.id) addFinding(findings, { level: "error", code: "self_dependency_milestone", message: `Milestone depends on itself: ${milestone.id}`, location: milestone.id });
        else if (resolved && !findMilestone(tree, resolved)) addFinding(findings, { level: "error", code: "missing_milestone_dependency", message: `Milestone dependency not found: ${dep}`, location: milestone.id });
      }

      for (const epic of milestone.epics) {
        if (!EPIC_ID_RE.test(epic.id)) addFinding(findings, { level: "error", code: "invalid_epic_id", message: `Invalid epic ID format: ${epic.id}`, location: epic.id });
        if (epicIds.has(epic.id)) addFinding(findings, { level: "error", code: "duplicate_epic_id", message: `Duplicate epic ID: ${epic.id}`, location: epic.id });
        epicIds.add(epic.id);

        for (const dep of epic.dependsOn) {
          const resolved = resolveEpicDepId(dep, milestone.id);
          if (resolved === epic.id) addFinding(findings, { level: "error", code: "self_dependency_epic", message: `Epic depends on itself: ${epic.id}`, location: epic.id });
          else if (resolved && !findEpic(tree, resolved)) addFinding(findings, { level: "error", code: "missing_epic_dependency", message: `Epic dependency not found: ${dep}`, location: epic.id });
        }

        for (const task of epic.tasks) {
          if (!TASK_ID_RE.test(task.id)) addFinding(findings, { level: "error", code: "invalid_task_id", message: `Invalid task ID format: ${task.id}`, location: task.id });
          if (taskIds.has(task.id)) addFinding(findings, { level: "error", code: "duplicate_task_id", message: `Duplicate task ID: ${task.id}`, location: task.id });
          taskIds.add(task.id);

          if (task.estimateHours === 0) {
            addFinding(findings, {
              level: "warning",
              code: "zero_estimate_hours",
              message: `Task estimate must be positive, got 0: ${task.id}`,
              location: task.id,
            });
          }

          for (const dep of task.dependsOn) {
            if (dep === task.id) addFinding(findings, { level: "error", code: "self_dependency_task", message: `Task depends on itself: ${task.id}`, location: task.id });
            else if (!allTasks.some((t) => t.id === dep)) {
              const resolvedEpic = resolveEpicDepId(dep, milestone.id);
              if (resolvedEpic && findEpic(tree, resolvedEpic)) {
                continue;
              }
              addFinding(findings, { level: "error", code: "missing_task_dependency", message: `Task dependency not found: ${dep}`, location: task.id });
            }
          }
        }
      }
    }
  }
}

function validateBugs(tree: TaskTree, findings: Finding[]): void {
  const dataDir = getDataDirName();
  const allTasks = getAllTasks(tree);
  const allIds = new Set(allTasks.map((t) => t.id));
  const bugIds = new Set<string>();

  for (const bug of (tree.bugs ?? [])) {
    if (!BUG_ID_RE.test(bug.id)) {
      addFinding(findings, { level: "error", code: "invalid_bug_id", message: `Invalid bug ID format: ${bug.id}`, location: bug.id });
    }
    if (bugIds.has(bug.id)) {
      addFinding(findings, { level: "error", code: "duplicate_bug_id", message: `Duplicate bug ID: ${bug.id}`, location: bug.id });
    }
    bugIds.add(bug.id);

    if (isTaskFileMissing(bug)) {
      addFinding(findings, { level: "error", code: "missing_bug_file", message: `Bug file missing for ${bug.id}`, location: `${dataDir}/${bug.file}` });
    }

    if (bug.estimateHours === 0) {
      addFinding(findings, { level: "warning", code: "zero_estimate_hours", message: `Bug estimate must be positive, got 0: ${bug.id}`, location: bug.id });
    }

    for (const dep of bug.dependsOn) {
      if (dep === bug.id) {
        addFinding(findings, { level: "error", code: "self_dependency_bug", message: `Bug depends on itself: ${bug.id}`, location: bug.id });
      } else if (!allIds.has(dep)) {
        addFinding(findings, { level: "error", code: "missing_bug_dependency", message: `Bug dependency not found: ${dep}`, location: bug.id });
      }
    }
  }
}

function validateCycles(tree: TaskTree, findings: Finding[]): void {
  const allTasks = getAllTasks(tree);
  const taskNodes = allTasks.map((t) => t.id);
  const taskEdges: Array<[string, string]> = [];
  for (const phase of tree.phases) {
    for (const milestone of phase.milestones) {
      for (const epic of milestone.epics) {
        epic.tasks.forEach((task, idx) => {
          for (const dep of task.dependsOn) {
            if (taskNodes.includes(dep)) {
              taskEdges.push([dep, task.id]);
            } else {
              const depEpic = findEpic(tree, resolveEpicDepId(dep, milestone.id));
              if (depEpic) {
                for (const depTask of depEpic.tasks) {
                  taskEdges.push([depTask.id, task.id]);
                }
              }
            }
          }
          if (task.dependsOn.length === 0 && idx > 0) taskEdges.push([epic.tasks[idx - 1]!.id, task.id]);
        });
      }
    }
  }
  // Include bugs in cycle detection
  for (const bug of (tree.bugs ?? [])) {
    for (const dep of bug.dependsOn) {
      if (taskNodes.includes(dep)) taskEdges.push([dep, bug.id]);
    }
  }
  const taskCycle = hasDirectedCycle(taskNodes, taskEdges);
  if (taskCycle) {
    addFinding(findings, {
      level: "error",
      code: "task_dependency_cycle",
      message: `Task dependency cycle detected: ${taskCycle.join(" -> ")}`,
      location: "task_dependencies",
    });
  }
}

async function validateRuntimeFiles(tree: TaskTree, findings: Finding[]): Promise<void> {
  const dataDir = getDataDirName();
  const contextPath = `${dataDir}/.context.yaml`;
  const sessionsPath = `${dataDir}/.sessions.yaml`;
  if (existsSync(contextPath)) {
    try {
      const context = (parse(await readFile(contextPath, "utf8")) as Record<string, unknown>) ?? {};
      const referenced = [
        context.current_task,
        context.primary_task,
        ...(((context.additional_tasks as string[]) ?? []) as string[]),
        ...(((context.sibling_tasks as string[]) ?? []) as string[]),
      ].filter(Boolean) as string[];
      for (const id of referenced) {
        if (!findTask(tree, id)) {
          addFinding(findings, { level: "warning", code: "stale_context_task", message: `Context references missing task: ${id}`, location: contextPath });
        }
      }
    } catch (err) {
      addFinding(findings, { level: "warning", code: "context_parse_error", message: `Could not parse context file: ${String(err)}`, location: contextPath });
    }
  }

  if (existsSync(sessionsPath)) {
    try {
      const sessions = (parse(await readFile(sessionsPath, "utf8")) as Record<string, unknown>) ?? {};
      for (const value of Object.values(sessions)) {
        if (!value || typeof value !== "object") continue;
        const taskId = (value as Record<string, unknown>).current_task as string | undefined;
        if (taskId && !findTask(tree, taskId)) {
          addFinding(findings, { level: "warning", code: "stale_session_task", message: `Session references missing task: ${taskId}`, location: sessionsPath });
        }
      }
    } catch (err) {
      addFinding(findings, { level: "warning", code: "sessions_parse_error", message: `Could not parse sessions file: ${String(err)}`, location: sessionsPath });
    }
  }
}

async function validateUninitializedTodos(tree: TaskTree, findings: Finding[]): Promise<void> {
  const dataDir = getDataDirName();
  const defaultMarkers = [
    "- TODO: Add requirements",
    "- TODO: Add acceptance criteria",
    "TODO: Add steps",
    "TODO: Describe expected behavior",
    "TODO: Describe actual behavior",
  ];

  const allTasks = getAllTasks(tree);
  for (const task of allTasks) {
    const taskFile = `${dataDir}/${task.file}`;
    if (!existsSync(taskFile)) {
      // Skip missing files - already reported by validateTreeFiles
      continue;
    }

    try {
      const content = await readFile(taskFile, "utf8");
      const hasDefaultContent = defaultMarkers.some((marker) => content.includes(marker));

      if (hasDefaultContent) {
        addFinding(findings, {
          level: "warning",
          code: "uninitialized_todo",
          message: `Task file still contains default placeholder content: ${task.id}`,
          location: taskFile,
        });
      }
    } catch (err) {
      // If we can't read the file, skip it
    }
  }
}

export async function runChecks(tasksDir?: string): Promise<{ ok: boolean; errors: Finding[]; warnings: Finding[]; summary: { errors: number; warnings: number; total: number } }> {
  const findings: Finding[] = [];
  const loader = new TaskLoader(tasksDir);
  const tree = await loader.load();
  validateTreeFiles(tree, findings);
  validateIdsAndDependencies(tree, findings);
  validateBugs(tree, findings);
  validateCycles(tree, findings);
  await validateRuntimeFiles(tree, findings);
  await validateUninitializedTodos(tree, findings);

  const errors = findings.filter((f) => f.level === "error");
  const warnings = findings.filter((f) => f.level === "warning");
  return {
    ok: errors.length === 0,
    errors,
    warnings,
    summary: { errors: errors.length, warnings: warnings.length, total: findings.length },
  };
}
